File: BST/I.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, node: Node, x):
        if node is None:
            return Node(x)

        if x < node.data:
            node.left = self.insert(node.left, x)
        else:
            node.right = self.insert(node.right, x)

        return node

    def height(self, node: Node):
        if node is None:
            return 0
        return 1 + max(self.height(node.left), self.height(node.right))

    def print_leaves(self, node: Node):
        if node is None:
            return

        self.print_leaves(node.left)
        if node.left is None and node.right is None:
            print(node.data)

        self.print_leaves(node.right)

    def is_balanced(self, node: Node):
        if node is None:
            return True

        lh = self.height(node.left)
        rh = self.height(node.right)

        if (abs(lh - rh) <= 1 and self.is_balanced(node.left) and self.is_balanced(node.right)):
            return True

        return False


a = list(map(int, input().split()))
a.pop()

visited = set()
tree = BinaryTree()
for item in a:
    if item not in visited:
        tree.root = tree.insert(tree.root, item)
        visited.add(item)

print("YES" if tree.is_balanced(tree.root) else "NO")

File: BST/I.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <string>

using namespace std;

struct Node
{
    int data;
    Node *left;
    Node *right;

    Node(int data) : data(data), left(nullptr), right(nullptr) {}
};

class BinaryTree
{
public:
    Node *root;

    BinaryTree() : root(nullptr) {}

    Node *insert(Node *node, int x)
    {
        if (node == nullptr)
        {
            return new Node(x);
        }

        if (x < node->data)
        {
            node->left = insert(node->left, x);
        }
        else
        {
            node->right = insert(node->right, x);
        }

        return node;
    }

    int height(Node *node)
    {
        if (node == nullptr)
            return 0;
        return 1 + max(this->height(node->left), this->height(node->right));
    }

    bool isBalanced(Node *node)
    {
        if (node == nullptr)
            return true;

        int lh = this->height(node->left);
        int rh = this->height(node->right);

        if (abs(lh - rh) <= 1 && this->isBalanced(node->left) && this->isBalanced(node->right))
            return true;

        return false;
    }

    void print_leaves(Node *node)
    {
        if (node == nullptr)
            return;

        print_leaves(node->left);
        if (node->left == nullptr && node->right != nullptr || node->left != nullptr && node->right == nullptr)
            cout << node->data << endl;

        print_leaves(node->right);
    }
};

int main()
{
    vector<int> a;
    int input;
    cin >> input;
    while (input != 0)
    {
        a.push_back(input);
        cin >> input;
    }

    set<int> visited;
    BinaryTree tree;
    for (int item : a)
    {
        if (visited.find(item) == visited.end())
        {
            tree.root = tree.insert(tree.root, item);
            visited.insert(item);
        }
    }

    cout << (tree.isBalanced(tree.root) ? "YES" : "NO");

    return 0;
}